WireAPI/Sources/WireAPI/APIs/TeamsAPI/TeamsAPIV0.swift

```diff
@@ -0,0 +1,289 @@
+//
+// Wire
+// Copyright (C) 2024 Wire Swiss GmbH
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see http://www.gnu.org/licenses/.
+//
+
+import Foundation
+
+class TeamsAPIV0: TeamsAPI {
+
+    let httpClient: HTTPClient
+    let decoder = ResponsePayloadDecoder(decoder: .defaultDecoder)
+
+    init(httpClient: HTTPClient) {
+        self.httpClient = httpClient
+    }
+
+    var apiVersion: APIVersion {
+        .v0
+    }
+
+    func basePath(for teamID: Team.ID) -> String {
+        switch apiVersion {
+        case .v0:
+            "/teams/\(teamID.transportString())"
+        default:
+            "/v\(apiVersion.rawValue)/teams/\(teamID.transportString())"
+        }
+    }
```

Comment:
yeah so currently `ZMTransportRequest` is adding this prefix, but it didn't feel right that the request should mutate the path, so I thought it'd be better to pass in the correct path when creating the request.

Of course though this means that every request for a different api version is different which would require every single method of the `TeamsAPI` to be overridden, even if no other changes to the API were present.

To try strike a balance between being explicit and avoiding duplication, I thought to have this inherited base path function.

While implementing this API I found that even a small change to an endpoint means to override the whole method and have some duplication. Sometimes I was wishing that I could only override part of the implementation, such as the request only, keeping the existing response handling, or vice versa. I'm trying to imagine some kind of hybrid solution where we have versioned API implementations but those can be composed from smaller units like Endpoint request generators and response parsers.

On the other hand, I also think the duplication that comes from more explicit code can also be tolerable given that duplication is minimal and not noisy with boilerplate code. That what I've tried to achieve with the `ResponseParser`, but I also was researching to see if other swift language features like result builder or macros could help here.

Reply:
I am not sure if I understand the issue, should we have a chat today after standup?

Do you mean because of an implementation like this, it uses `basePath(for: teamID)`? From my understanding it should be resolved correctly in subclasses if `basePath` has been overridden, because we have dynamic dispatch for each method.
```swift
func getTeam(for teamID: Team.ID) async throws -> Team {
        let request = HTTPRequest(
            path: basePath(for: teamID),
            method: .get
        )
```

Reply:
done 4e4840800f873dc3ad743af161c93d82b25a2727